plugins/saveactions.c

```diff
@@ -258,49 +271,387 @@ static void backupcopy_document_save_cb(GObject *obj, GeanyDocument *doc, gpoint
+	document_save_file(doc, TRUE);
+}
+
+
 static void instantsave_document_new_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
 {
-    if (enable_instantsave && doc->file_name == NULL)
-    {
-		const gchar *directory;
-		gchar *new_filename;
-		gint fd;
-		GeanyFiletype *ft = doc->file_type;
-
-		if (ft == NULL || ft->id == GEANY_FILETYPES_NONE)
-			/* ft is NULL when a new file without template was opened, so use the
-			 * configured default file type */
-			ft = filetypes_lookup_by_name(instantsave_default_ft);
-
-		/* construct filename */
-		directory = !EMPTY(instantsave_target_dir) ? instantsave_target_dir : g_get_tmp_dir();
-		new_filename = g_build_filename(directory, "gis_XXXXXX", NULL);
-		if (ft != NULL && !EMPTY(ft->extension))
-			SETPTR(new_filename, g_strconcat(new_filename, ".", ft->extension, NULL));
-
-		/* create new file */
-		fd = g_mkstemp(new_filename);
-		if (fd == -1)
+	if (enable_instantsave && doc->file_name == NULL)
+	{
+		save_new_document_to_disk(
+			doc,
+			!EMPTY(instantsave_target_dir) ? instantsave_target_dir : g_get_tmp_dir(),
+			_("gis_XXXXXX"),
+			instantsave_default_ft
+		);
+	}
+}
+
+
+static void persistent_temp_files_document_new_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	if (enable_persistent_temp_files && doc->file_name == NULL)
+	{
+		if (EMPTY(persistent_temp_files_target_dir))
+		{
+			dialogs_show_msgbox(GTK_MESSAGE_ERROR,
+				_("Persistent temp files directory does not exist or is not writable."));
+			return;
+		}
+
+		save_new_document_to_disk(
+			doc,
+			persistent_temp_files_target_dir,
+			_("gtemp_XXXXXX"),
+			filetypes[GEANY_FILETYPES_NONE]->name
+		);
+	}
+}
+
+
+static gboolean is_temp_saved_file(const gchar *filename)
+{
+	static const gchar *temp_saved_file_prefix = "gtemp_";
+
+	return g_str_has_prefix(filename, temp_saved_file_prefix);
+}
+
+
+static gboolean is_temp_saved_file_doc(GeanyDocument *doc)
+{
+	gchar *file_path, *filename;
+	gboolean matched;
+
+	file_path = doc->file_name;
+
+	if (file_path == NULL)
+	{
+		return FALSE;
+	}
+
+	filename = g_path_get_basename(file_path);
+	matched = is_temp_saved_file(filename);
+
+	g_free(filename);
+
+	return matched;
+}
+
+
+static void load_all_temp_files_into_editor(const gchar *path)
+{
+	GDir *dir;
+	GError *error = NULL;
+	gchar *utf8_filename, *locale_file_path;
+	const gchar *filename;
+
+	dir = g_dir_open(path, 0, &error);
+	if (dir == NULL)
+	{
+		dialogs_show_msgbox(GTK_MESSAGE_ERROR,
+						_("Persistent temp files directory not found"));
+		return;
+	}
+
+	foreach_dir(filename, dir)
+	{
+		utf8_filename = utils_get_utf8_from_locale(filename);
+		locale_file_path = g_build_path(G_DIR_SEPARATOR_S, path, filename, NULL);
+
+		if (is_temp_saved_file(utf8_filename))
+		{
+			document_open_file(locale_file_path, FALSE, NULL, NULL);
+		}
+
+		g_free(utf8_filename);
+		g_free(locale_file_path);
+	}
+
+	g_dir_close(dir);
+}
+
+
+static gboolean file_exists(gchar *utf8_file_path)
```

Comment:
Instead of this function you can just check `doc->real_path != NULL`.

Reply:
done